feat(shutdown): implement graceful shutdown for camera session

EthanDM opened the pull request:
## What
Fix the navigation freeze on React Navigation/Fabric by shutting the camera session down deterministically when the view unmounts.

## Changes
- Add CameraSession.shutdown, which disables photo/video/audio/code-scanner/location/torch before flipping isActive = false, and clears delegates while removing outputs.  
- Stop both capture sessions asynchronously during CameraSession.deinit as a backstop.  
- Have CameraView call the shutdown helper on unmount/deinit, log if teardown exceeds 2 s, and guard against concurrent shutdowns.

## Tested on
- iPhone 14 Pro (iOS 26.0.1) with code scanner enabled, navigating away from the camera screen (no freeze).

## Related issues
- Fixes #3636

EthanDM committed:
feat(shutdown): implement graceful shutdown for camera session

- Add shutdown method to stop all outputs and release resources
- Schedule shutdown with warning if it takes longer than expected

jkaufman commented:
@EthanDM Thank you for your effort on this. We’ve been testing [a patch](https://github.com/mrousavy/react-native-vision-camera/pull/3664) that builds on yours and would appreciate a look.

mrousavy commented on package/ios/Core/CameraSession.swift:
```diff
@@ -121,10 +136,12 @@ final class CameraSession: NSObject, AVCaptureVideoDataOutputSampleBufferDelegat
         try lambda(config)
       } catch CameraConfiguration.AbortThrow.abort {
         // call has been aborted and changes shall be discarded
+        completionBlock?()
```
instead of writing that 3 times, put it in a `defer { ... }` block so we dont forget it when we add a fourth exit point

mrousavy commented on package/ios/React/CameraView.swift:
```diff
@@ -283,6 +290,31 @@ public final class CameraView: UIView, CameraSessionDelegate, PreviewViewDelegat
+    CameraQueues.cameraQueue.asyncAfter(deadline: .now() + .seconds(2), execute: slowShutdownWarning)
+
+    cameraSession.shutdown { [weak self] in
+      slowShutdownWarning.cancel()
```
```suggestion
    let logSlowShutdownWarning = DispatchWorkItem {
      VisionLogger.log(level: .warning, message: "CameraSession shutdown is still running after 2 seconds.")
    }
    CameraQueues.cameraQueue.asyncAfter(deadline: .now() + .seconds(2), execute: logSlowShutdownWarning)

    cameraSession.shutdown { [weak self] in
      logSlowShutdownWarning.cancel()
```

mrousavy commented on package/ios/Core/CameraSession+Configuration.swift:
```diff
@@ -61,6 +61,12 @@ extension CameraSession {
+      }
+      if let videoOutput = output as? AVCaptureVideoDataOutput {
+        videoOutput.setSampleBufferDelegate(nil, queue: nil)
+      }
```
I'm 99% sure this is unnecessary. The outputs store the delegates as `weak` references, and there's no Thread hops involved.

Can you try to remove this code and check if it still works fine for you? I'd rather have that out tbh because it is super specific and feels like monkeypatching

mrousavy reviewed (changes requested):
Thanks for this PR - seems like there were some issues with proper session shutdown on Fabric - I wonder why that wasn't a problem before on Paper. Maybe view caching/recycling?

Anyways - I left some review comments

EthanDM committed:
Rename slowShutdownWarning to logSlowShutdownWarning

Co-authored-by: Marc Rousavy <me@mrousavy.com>

EthanDM committed:
fix(shutdown): resolve shutdown freeze issue on iOS 26

- Remove unnecessary delegate assignments before removing outputs
- Ensure completion block is called in all error scenarios during configuration

EthanDM commented:
Thanks for the review @mrousavy, I addressed all of the comments and retested. Still no freezes during mount/unmount.

Let me know if anything else is needed.

nakshatrasinghh commented:
This seems important can it be merged please?

mrousavy commented:
Hey - thanks for your PR, I really appreciate it! And sorry for not replying sooner, I was hyper-focused on [VisionCamera V5](https://github.com/mrousavy/react-native-vision-camera/releases/tag/v5.0.0).

I believe this is now obsolete, as VisionCamera V5 "fixed" this, I'm closing this PR (as it's out of sync with main since I rewrote the entire codebase), but please let me know if this still needs attention in V5 and we can open a new issue for discussion, or possibly a PR already with a fix.
